flask_sqlalchemy/__init__.py

```diff
@@ -730,19 +730,23 @@ class User(db.Model):
        naming conventions among other, non-trivial things.
     """
 
-    def __init__(self, app=None, use_native_unicode=True, session_options=None, metadata=None):
+    def __init__(self, app=None, use_native_unicode=True, session_options=None,
+                 metadata=None, query_class=BaseQuery, model_class=Model):
 
         if session_options is None:
             session_options = {}
 
         session_options.setdefault('scopefunc', connection_stack.__ident_func__)
+        session_options.setdefault('query_cls', query_class)
         self.use_native_unicode = use_native_unicode
         self.session = self.create_scoped_session(session_options)
-        self.Model = self.make_declarative_base(metadata)
-        self.Query = BaseQuery
+        self.Model = self.make_declarative_base(model_class, metadata)
+        # ugly hack, is there a better way?
+        self.Model.query_class = query_class
```

Comment:
I'm not happy with this line. I'd considered moving it to `make_declarative_base` but I'm not 100%. Should we respect the query class on the model, leave this as is/move to `make_declarative_base`, add a toggle for replacing the query class or something else?